### ISSUE TITLE
Build inside Docker fails with many "undefined reference to dl" errors

### ISSUE THREAD

**Issue by mykytak** [avast/retdec#623]
Got error while building from Dockerfile:

```
../../deps/openssl/openssl/src/openssl-install/lib/libcrypto.a(dso_dlfcn.o): In function `dlfcn_globallookup':
dso_dlfcn.c:(.text+0x11): undefined reference to `dlopen'
dso_dlfcn.c:(.text+0x24): undefined reference to `dlsym'
dso_dlfcn.c:(.text+0x2f): undefined reference to `dlclose'
../../deps/openssl/openssl/src/openssl-install/lib/libcrypto.a(dso_dlfcn.o): In function `dlfcn_bind_func':
dso_dlfcn.c:(.text+0x1c4): undefined reference to `dlsym'
dso_dlfcn.c:(.text+0x282): undefined reference to `dlerror'
../../deps/openssl/openssl/src/openssl-install/lib/libcrypto.a(dso_dlfcn.o): In function `dlfcn_load':
dso_dlfcn.c:(.text+0x2f1): undefined reference to `dlopen'
dso_dlfcn.c:(.text+0x369): undefined reference to `dlclose'
dso_dlfcn.c:(.text+0x3a5): undefined reference to `dlerror'
../../deps/openssl/openssl/src/openssl-install/lib/libcrypto.a(dso_dlfcn.o): In function `dlfcn_pathbyaddr':
dso_dlfcn.c:(.text+0x471): undefined reference to `dladdr'
dso_dlfcn.c:(.text+0x4d6): undefined reference to `dlerror'
../../deps/openssl/openssl/src/openssl-install/lib/libcrypto.a(dso_dlfcn.o): In function `dlfcn_unload':
dso_dlfcn.c:(.text+0x6b2): undefined reference to `dlclose'
collect2: error: ld returned 1 exit status
```

**Comment by silverbacknet** [comment]
Those are kernel functions that should always be available. Please add at a minimum your operating system and version, your compiler and version, and what version of retdec you're building.

**Comment by mykytak** [comment]
@silverbacknet I just copied dockerfile from repo and run it.

With prebuilded package (v3.3), got `.../retdec/bin/retdec-fileinfo: error while loading shared libraries: libcrypto.so.1.1: cannot open shared object file: No such file or directory` error. Which I think is the same error in the end.
Tried to solve by installing libssl-dev (both 64 and 32 bit versions), no luck.

Docker version 18.09.7, build 2d0083d
Ubuntu 16.04.6 LTS

**Comment by s3rvac** [comment]
Hi. Thank you for the report!

As for the `dl`-related errors that you get when building via Docker: I have managed to reproduce the issue by following these steps:
```
$ git clone https://github.com/avast/retdec
$ cd retdec
$ docker build -t retdec - < Dockerfile
```
There is a bug in our build system. I will let you know when the bug is fixed.

As for the OpenSSL-related errors that you get with [our 3.3 Ubuntu release](https://github.com/avast/retdec/releases/download/v3.3/retdec-v3.3-ubuntu-64b.zip): Our Ubuntu release requires OpenSSL 1.1, while Ubuntu 16.04 only provides [OpenSSL 1.0](https://packages.ubuntu.com/xenial/openssl). You will need to either upgrade to Ubuntu 18.04 or find a way of installing OpenSSL 1.1 into your Ubuntu 16.04.

**Comment by s3rvac** [comment]
There was a missing dependency of OpenSSL on `libdl`. I have submitted a fix in #624. After it is merged, builds via Docker will succeed.

Apart from the above fix, in commit 1fd3328a2705f69922b31e31780378f7ebce7ef6, I have modified `Dockerfile` to install `openssl` and `libssl-dev` from Ubuntu repositories. This will enable RetDec to use them instead of building OpenSSL by itself, which also solves the issues with builds inside Docker. Apart from that, it makes RetDec build faster inside Docker as it will no longer need to build OpenSSL (it will use a pre-built version). Please, try to use `Dockerfile` from the current `master`, and let us know if the build succeeds for you.

**Comment by PeterMatula** [comment]
Merged #624.

Also added 2 new TeamCity configurations `linux-dockerfile-build` and `linux-dockerfile-dev-build` that build `Dockerfile` and `Dockerfile.dev`. Builds are triggered only if relevant files are changed in the `master` branch. This should let us know if Docker files are broken in the future.

**Comment by mykytak** [comment]
Thanks @s3rvac ! For quick response and fix and clarifying about OpenSSL version. Docker version works for me now.
